src/Storages/MergeTree/RowOrderOptimizer.cpp

```diff
@@ -0,0 +1,162 @@
+///          ----------------------
+///          1      1     a     b
+///          1      1     b     e
+///          --------
+///          1      2     e     a
+///          1      2     d     c
+///          1      2     e     a
+///          --------
+///          2      1     a     3
+///          ----------------------
+EqualRanges getEqualRanges(const Block & block, const SortDescription & sort_description, const IColumn::Permutation & permutation)
+{
+    EqualRanges ranges;
+    const size_t rows = block.rows();
+    if (sort_description.empty())
+    {
+        ranges.push_back({0, rows});
+    }
+    else
+    {
+        for (size_t i = 0; i < rows;)
+        {
+            size_t j = i;
+            while (j < rows && haveEqualSortingKeyValues(block, sort_description, permutation[i], permutation[j]))
+                ++j;
+            ranges.push_back({i, j});
+            i = j;
+        }
+    }
+    return ranges;
+}
+
+std::vector<size_t> getCardinalitiesInPermutedRange(
+    const Block & block,
+    const std::vector<size_t> & other_column_indexes,
+    const IColumn::Permutation & permutation,
+    const EqualRange & equal_range)
+{
+    std::vector<size_t> cardinalities(other_column_indexes.size());
+    for (size_t i = 0; i < other_column_indexes.size(); ++i)
+    {
+        const ColumnPtr & column = block.getByPosition(i).column;
+        cardinalities[i] = column->estimateCardinalityInPermutedRange(permutation, equal_range);
+    }
+    return cardinalities;
+}
+
+void updatePermutationInEqualRange(
+    const Block & block,
+    const std::vector<size_t> & other_column_indexes,
+    IColumn::Permutation & permutation,
+    const EqualRange & equal_range,
+    const std::vector<size_t> & cardinalities)
+{
+    std::vector<size_t> column_order(other_column_indexes.size());
+    iota(column_order.begin(), column_order.end(), 0);
+    auto cmp = [&](size_t lhs, size_t rhs) -> bool { return cardinalities[lhs] < cardinalities[rhs]; };
+    ::sort(column_order.begin(), column_order.end(), cmp);
+
+    std::vector<EqualRange> ranges = {equal_range};
+    for (size_t i : column_order)
+    {
+        const size_t column_id = other_column_indexes[i];
+        const ColumnPtr & column = block.getByPosition(column_id).column;
+        column->updatePermutation(IColumn::PermutationSortDirection::Ascending, IColumn::PermutationSortStability::Unstable, 0, 1, permutation, ranges);
+    }
+}
+
+}
+
+void RowOrderOptimizer::optimize(const Block & block, const SortDescription & description, IColumn::Permutation & permutation)
+{
+    if (block.columns() == 0)
+        return; /// a table without columns, this should not happen in the first place ...
+
+    if (permutation.empty())
+    {
+        const size_t rows = block.rows();
+        permutation.resize(rows);
+        iota(permutation.data(), rows, IColumn::Permutation::value_type(0));
+    }
+
+    const EqualRanges equal_ranges = getEqualRanges(block, description, permutation);
```

Comment:
To simplify investigation/debugging, we should have some Trace-level logging (src/Common/logger_useful.h), i.e. `LOG_TRACE` in this file which prints basic statistics such as the number of equal ranges and cardinalities.